STM32N6 : Add the support of the USB and UDC

marwaiehm-st opened the pull request:
Add the support of the USB OTG HS
Enable the USB OTG HS node of the stm32n6570_dk and nucleo_n657x0_q boards.

marwaiehm-st committed:
dts: bindings: usb:  Introduce STM32N6 OTG HS compatible

Add STM32N6 USB OTG HS compatible 'st,stm32n6-otghs'
that doesn't require pinctrl-0 and pinctrl-names.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
boards: st: nucleo_n657x0_q: Enable OTG HS peripheral

Enable the USB OTG HS node of the nucleo_n657x0_q board

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
dts: arm: STM32N6X serie with OTG HS instance

Add the USB OTG HS node for the STM32N6X devices

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
boards: st: stm32n6570_dk: Enable OTG HS peripheral

Enable the USB OTG HS node of the stm32n6570_dk board

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
drivers: usb: Add STM32N6 USB support

Enable clock and power for the OTG HS peripheral
of the STM32N6x serie

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

mbolivar commented on soc/st/stm32/stm32n6x/CMakeLists.txt:
```diff
@@ -15,13 +15,25 @@ else()
+  Signing Image tool (${SIGNING_TOOL}) is not available.
+  Signed image will not be generated.
+  You won't be able to run image on board.")
+endif()
+
+if(NOT SIGNING_TOOL_FIND STREQUAL SIGNING_TOOL_FIND-NOTFOUND)
```
else()

mbolivar commented on dts/bindings/usb/st,stm32n6-otghs.yaml:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2024, STMicroelectronics
```
2025?

Same question for the rest of the bindings changes.

marwaiehm-st replied:
Done

mbolivar reviewed (commented):
Runner changes look reasonable to me, no objections. Found a couple of nonblocking minor things

gautierg-st commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -402,13 +416,27 @@ static int usb_dc_stm32_clock_enable(void)
+	USB1_HS_PHYC->USBPHYC_CR |= USB_USBPHYC_CR_FSEL_24MHZ;
+
+	/* Peripheral OTGPHY clock enable */
+	LL_AHB5_GRP1_EnableClock(LL_AHB5_GRP1_PERIPH_OTGPHY1);
```
Can't you use `clock_control_on` here?

gautierg-st commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -424,12 +452,14 @@ static int usb_dc_stm32_clock_disable(void)
 #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
 	clock_control_off(clk, (clock_control_subsys_t)&phy_pclken[0]);
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	LL_AHB5_GRP1_DisableClock(LL_AHB5_GRP1_PERIPH_OTGPHY1);
```
Ditto with `clock_control_off`

gautierg-st commented on west.yml:
```diff
@@ -238,7 +238,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 1d1f81866ccbaa6e84e9960ed763e005d1e45560
+      revision: ea773e786a9f78d779480b5cec8c3c98828f2127
```
This is not the latest version of the HAL. But you'll be able to drop this commit entirely when #84985 is merged.

marwaiehm-st replied:
Done, I changed the revision

marwaiehm-st committed:
boards: st: stm32n6: Add the usbd test feature

Add the usbd test feature to twister.yaml for the boards
nucleo_n657x0_q and stm32n6570_dk.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -481,11 +493,7 @@ static int usb_dc_stm32_init(void)
-#else
-	usb_dc_stm32_state.pcd.Instance = USB_OTG_FS;
-#endif
+	usb_dc_stm32_state.pcd.Instance = (USB_OTG_GlobalTypeDef *)USB_BASE_ADDRESS;
```
Out of the scope of this, PR, but this could be extended to above lines:
```
#ifdef USB
	usb_dc_stm32_state.pcd.Instance = USB;
#else
	usb_dc_stm32_state.pcd.Instance = USB_DRD_FS;
#endif
```
And same approach can be used on usb next driver.

erwango replied:
cc: @kbidani

marwaiehm-st replied:
So, it should be reverted and pushed in another PR

erwango replied:
No no. Let's keep as is. But this should actually be extended in follow up PR.

marwaiehm-st replied:
ok

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -402,13 +412,15 @@ static int usb_dc_stm32_clock_enable(void)
 
 #if USB_OTG_HS_EMB_PHY
+#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_otghs)
 	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_OTGPHYC);
```
Out of the scope of this PR, but we need to understand where this line is coming from and configure clock using clock_contol API + correct dts description

erwango replied:
cc: @kbidani

erwango commented on west.yml:
```diff
@@ -238,7 +238,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 1d1f81866ccbaa6e84e9960ed763e005d1e45560
+      revision: 55043bcc35fffa3b4a8c75a696d932b5020aad09
```
Since https://github.com/zephyrproject-rtos/zephyr/pull/84985#event-16213061785 was merged, this can be dropped.

jfischer-no commented on dts/bindings/usb/st,stm32n6-otghs.yaml:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2025, STMicroelectronics
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  STM32N6 OTGHS controller
+
+  In the STM32N6 series, the `pinctrl-0` and `pinctrl-names` properties are not required
+  for the USB OTG HS peripheral configuration. This is because the pin multiplexing
+  for the USB OTG HS peripheral are handled automatically by the hardware.
+
+compatible: "st,stm32n6-otghs"
+
+include: st,stm32-otghs-common.yaml
```
@marwaiehm-st @erwango This file is not really necessary. You can just use `compatible = "st,stm32n6-otghs", "st,stm32-otghs";`. In the driver, use first compatible to implement a hook to handle this particular type, what would make your code more maintainable.

erwango replied:
> This file is not really necessary. 

Are you sure ?
Since this particular IP doesn't make use of pinctrl, it is correct to specify a new compatible that doesn't make `pinctrl` prop required. We can think about making `pinctrl` prop not required in ` "st,stm32-otghs"` compat, but this will reduce the ability to catch wrong configurations and anyway a `"st,stm32n6-otghs"` compat is required to handle IP specifics as you mentioned..

Unless I missed something...

jfischer-no replied:
> Are you sure ?

Yes, that is how we handle vendor specific hooks in DWC2 driver. There is zero noise in bindings or vendor/type independent code.

erwango replied:
I'm not sure to understand. You did provide a [`dts/bindings/usb/st,stm32f4-fsotg.yaml`](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/usb/st%2Cstm32f4-fsotg.yaml) for `STM32F4 SoC series OTG_FS DWC2 compatible controller` variant.

jfischer-no replied:
This is to be closer to the scheme the Linux kernel uses for the DWC2 driver.
My point is that something like dts/bindings/usb/st,stm32n6-otghs.yaml without any properties does not make sense, even in the code it looks like a hack

```c
#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_otghs)
PINCTRL_DT_INST_DEFINE(0);
static const struct pinctrl_dev_config *usb_pcfg =
					PINCTRL_DT_INST_DEV_CONFIG_GET(0);
#endif
```

 I suggest to remove required from the pinctl properties.

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -37,6 +37,7 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #define UDC_STM32_IRQ_NAME     usb
 #endif
 
+#define USB_BASE_ADDRESS	DT_INST_REG_ADDR(0)
```
The following two lines of code use the UDC_STM32_ prefix. Apart from that this is not the correct way and these properties should be stored in the driver config to support multiple instances, why not follow the same schema and name macro UDC_STM32_BASE?

marwaiehm-st replied:
I renamed the macro: UDC_STM32_BASE_ADDRESS

marwaiehm-st replied:
I removed the cleanup modifications because it's already addressed in another PR: https://github.com/zephyrproject-rtos/zephyr/pull/86100

jfischer-no commented on boards/st/stm32n6570_dk/twister.yaml:
```diff
@@ -17,6 +17,7 @@ supported:
   - spi
   - uart
   - usb_device
+  - usbd
```
Please squash it with the "boards: st: nucleo_n657x0_q: Add the usbd" commit.

marwaiehm-st replied:
Done

marwaiehm-st committed:
drivers: usb: Add STM32N6 family support to UDC driver

Add STM32N6 family support to UDC driver

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

jfischer-no reviewed (dismissed):
I will not be asking for any changes, just some notes for future commits/PRs.
The USB device controller API is UDC, which is what USB device stack next uses, but USB stack does not really matter here in the drivers. The commit message should be like `driver: usb: add STM32n6 family support to UDC driver`.

There is nothing like `usb_next`.

"Proper capitalization of nouns" should be used, e.g. `drivers: usb: Add STM32N6 USB support`.

`usbd` is a test feature,  `boards: st: stm32n6: Add the usbd test feature`

kartben commented on boards/st/stm32n6570_dk/doc/index.rst:
```diff
@@ -66,6 +66,36 @@ Supported Features
++--------------------------+------------------------------+
+| OTG2_TXRTUNE     | USB OTG2 Transmit Retune             |
++---------+-----------------+-----------------------------+
+
```
formatting of this table is broken

kartben commented on boards/st/stm32n6570_dk/doc/index.rst:
```diff
@@ -66,6 +66,36 @@ Supported Features
 
 .. zephyr:board-supported-hw::
 
+USB:
+====
```
```suggestion
USB
===
```

jfischer-no commented on dts/bindings/usb/st,stm32-otghs.yaml:
This is Zephyr specific, right? @mbolivar How will property-blocklist/allowlist be handled in the future?

marwaiehm-st commented:
I added the label DNM because we should wait for the PR https://github.com/zephyrproject-rtos/zephyr/pull/86100 to be merged first

marwaiehm-st committed:
drivers: usb: Enhance USB and UDC drivers for STM32 OTG configurations

- Unified the handling of USB OTG HS and USB OTG FS
  by removing redundant preprocessor conditionals.
- Introduced a new macro `UDC_STM32_BASE_ADDRESS`
  to dynamically set the USB base address.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st commented:
I added a commit to read the correct USB OTG configurations (instance), so there is no need to wait for this PR https://github.com/zephyrproject-rtos/zephyr/pull/86100 to be merged anymore.
